cpp/test/sg/fil_test.cu

```diff
@@ -1069,4 +1072,99 @@ TEST_P(TreeliteThrowSparse8FilTest, Import) { check(); }
 INSTANTIATE_TEST_CASE_P(FilTests, TreeliteThrowSparse8FilTest,
                         testing::ValuesIn(import_throw_sparse8_inputs));
 
+template <typename T>
+__device__ void serial_multireduction(const T* in, T* out, int set_size,
+                                      int n_sets) {
+  __syncthreads();
+  if (threadIdx.x < set_size) {
+    int reduction_id = threadIdx.x;
+    T sum{};
+    for (int set = 0; set < n_sets; ++set)
+      sum += in[reduction_id + set * set_size];
+    out[reduction_id] = sum;
+  }
+  __syncthreads();
+}
+
+// the most threads a block can have
+const int max_threads = 1024;
+
+template <int R, typename T>
+__device__ void test_single_radix(const T value, const int valid_threads) {
+  __shared__ T work[max_threads], correct_result[max_threads];
+  for (int set_size = 1; set_size < valid_threads; ++set_size) {
+    int n_sets = valid_threads / set_size;
+    work[threadIdx.x] = value;
+    serial_multireduction(work, correct_result, set_size, n_sets);
+    T sum = multireduction<R>(work, set_size, n_sets);
+    if (threadIdx.x < set_size && sum != correct_result[threadIdx.x]) {
+      printf("multireduction<%d>(on %d sets sized %d)[%d] gave wrong result\n",
+             R, n_sets, set_size, threadIdx.x);
+      asm("trap;");
```

Comment:
TODO: test that this will cause `CUDA_CHECK(cudaDeviceSynchronize())` to fail the test